mlir/test/mlir-pdll/CodeGen/MLIR/expr.pdll

```diff
@@ -142,3 +142,99 @@ Pattern RangeExpr {
 // CHECK: %[[TYPE:.*]] = type : i32
 // CHECK: operation({{.*}}) -> (%[[TYPE]] : !pdl.type)
 Pattern TypeExpr => erase op<> -> (type<"i32">);
+
+// -----
+
+//===----------------------------------------------------------------------===//
+// Parse attributes and rewrite
+//===----------------------------------------------------------------------===//
+
+// Rewriter helpers declarations.
+Rewrite createDictionaryAttr() -> Attr;
+Rewrite addEntryToDictionaryAttr(dict: Attr, attrName: Attr, attr : Attr) -> Attr;
+Rewrite createArrayAttr() -> Attr;
+Rewrite addElemToArrayAttr(arrayAttr: Attr, newElement: Attr) -> Attr;
```

Comment:
I think the test cases should be as self contained as possible. Let's have `// -----` in between them 
and add only the relevant native calls to each test case